File: app/src/main/java/br/com/jonathanzanella/myexpenses/account/transactions/MonthTransactionsView.java
package br.com.jonathanzanella.myexpenses.account.transactions;

import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.widget.TextView;

import org.joda.time.DateTime;

import java.text.SimpleDateFormat;
import java.util.Locale;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.views.BaseView;
import butterknife.BindView;
import butterknife.ButterKnife;

public class MonthTransactionsView extends BaseView implements MonthTransactionsContractView {
	public final SimpleDateFormat simpleDateFormat = new SimpleDateFormat("MMMM/yy", Locale.getDefault());

	@BindView(R.id.view_month_transactions_list)
	RecyclerView list;
	@BindView(R.id.view_month_transactions_header)
	TextView header;
	@BindView(R.id.view_month_transactions_balance)
	TextView balance;

	int singleRowHeight;

	String monthTransactionsTemplate;

	private MonthTransactionsPresenter presenter;
	private LoadTransactionsCallback loadTransactionsCallback;

	public MonthTransactionsView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public MonthTransactionsView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	@Override
	protected void init() {
		singleRowHeight = getContext().getResources().getDimensionPixelSize(R.dimen.single_row_height);

		monthTransactionsTemplate = getResources().getString(R.string.month_transactions);
		inflate(getContext(), R.layout.view_account_month_transactions, this);
		ButterKnife.bind(this);
		presenter = new MonthTransactionsPresenter(getContext(), this);

		list.setAdapter(presenter.getAdapter());
		list.setHasFixedSize(true);
		list.setLayoutManager(new LinearLayoutManager(getContext()));
		list.setNestedScrollingEnabled(false);
	}

	void showBalance(final Account account, final DateTime month, int balance) {
		header.setText(monthTransactionsTemplate.concat(" ").concat(simpleDateFormat.format(month.toDate())));
		presenter.showBalance(account, month, balance);
	}

	@Override
	public void onBalanceUpdated(int balance) {
		this.balance.setText(CurrencyHelper.format(balance));
		//noinspection deprecation
		this.balance.setTextColor(getResources().getColor(balance >= 0 ? R.color.value_unreceived : R.color.value_unpaid));

		if(loadTransactionsCallback != null)
			loadTransactionsCallback.onTransactionsLoaded(balance);

		list.setMinimumHeight(list.getAdapter().getItemCount() * singleRowHeight);
	}

	public void setLoadTransactionsCallback(LoadTransactionsCallback loadTransactionsCallback) {
		this.loadTransactionsCallback = loadTransactionsCallback;
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/account/transactions/MonthTransactionsView.kt
package br.com.jonathanzanella.myexpenses.account.transactions

import android.content.Context
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.LinearLayoutManager
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.extensions.fromIoToMainThread
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.transaction.TransactionAdapter
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.view_account_month_transactions.view.*
import org.joda.time.DateTime
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

class MonthTransactionsView@JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), MonthTransactionsContractView {

    private val adapter = TransactionAdapter()
    private val simpleDateFormat = SimpleDateFormat("MMMM/yy", Locale.getDefault())
    private var singleRowHeight = context.resources.getDimensionPixelSize(R.dimen.single_row_height)
    private val monthTransactionsTemplate = resources.getString(R.string.month_transactions)
    private var showBalanceDisposable: Disposable? = null

    @Inject
    lateinit var presenter: MonthTransactionsPresenter

    init {
        App.getAppComponent().inject(this)
    }

    init {
        View.inflate(context, R.layout.view_account_month_transactions, this)

        list.adapter = adapter
        list.setHasFixedSize(true)
        list.layoutManager = LinearLayoutManager(context)
        list.isNestedScrollingEnabled = false
    }

    internal fun showBalance(account: Account, month: DateTime) {
        showBalanceDisposable?.dispose()

        val date = simpleDateFormat.format(month.toDate())
        header.text = "$monthTransactionsTemplate $date"

        showBalanceDisposable = presenter.getAccountTransactions(account, month)
                .doOnNext { adapter.setTransactions(it) }
                .flatMap { presenter.calculateAccountBalance(account, it) }
                .fromIoToMainThread()
                .subscribe({
                    adapter.notifyDataSetChanged()
                    onBalanceUpdated(it)
                }, { e ->
                    Timber.tag(javaClass.canonicalName).e("Error showing transactions, message ${e.message}")
                })
    }

    override fun onDetachedFromWindow() {
        showBalanceDisposable?.dispose()
        super.onDetachedFromWindow()
    }

    override fun onBalanceUpdated(balance: Int) {
        this.balance.text = balance.toCurrencyFormatted()
        this.balance.setTextColor(ResourcesCompat.getColor(resources, (if (balance >= 0) R.color.value_unreceived else R.color.value_unpaid), null))

        list.minimumHeight = list.adapter.itemCount * singleRowHeight
    }
}
